### ISSUE TITLE
Proposal: Bind deis containers to internal IPs

### ISSUE THREAD

**Issue by ineu** [deis/deis#3168]
As deis assumes that cluster is connected to internal, private network, it should not expose its ports to the public.
For example, https://github.com/deis/deis/blob/master/deisctl/units/deis-database.service#L10. Here
```
ExecStart=/bin/sh -c "IMAGE=`/run/deis/bin/get_image /deis/database` && docker run --name deis-database --volumes-from=deis-database-data -p 5432:5432 -e EXTERNAL_PORT=5432 -e HOST=$COREOS_PRIVATE_IPV4 $IMAGE"
```
could be changed to something like 
```
ExecStart=/bin/sh -c "IMAGE=`/run/deis/bin/get_image /deis/database` && docker run --name deis-database --volumes-from=deis-database-data -p $COREOS_PRIVATE_IPV4:5432:5432 -e EXTERNAL_PORT=5432 -e HOST=$COREOS_PRIVATE_IPV4 $IMAGE"
```
This would not only make things more secure, as services like deis-database should not be accessed from outside, but would also allow to start attached services in the same cluster on usual ports.

**Comment by jannispl** [comment]
I don't agree with this, as I am looking forward to connecting nodes from multiple data centers together which don't always provide me with a private network. I won't get around using the public interface and setting up a custom firewall.

**Comment by aledbf** [comment]
@ineu deis includes a firewall script to help with the security in the cluster http://docs.deis.io/en/latest/managing_deis/security_considerations.html?highlight=firewall#running-deis-on-public-clouds

**Comment by ineu** [comment]
@jannispl This is system requirement:
http://docs.deis.io/en/latest/installing_deis/system-requirements.html#network

> Due to changes introduced in Docker 1.3.1 related to insecure Docker registries, the hosts running Deis must be able to communicate via a private network in one of the RFC 1918 or RFC 6598 private address spaces: 10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, or 100.64.0.0/10.

**Comment by ineu** [comment]
@aledbf I've tried it and it didn't really work for me: docker ACCEPT rules were inserted before firewall rules, so it didn't block anything. Maybe I've used it wrong, need to debug.
But the real problem is that Deis services utilize ports that should be free. It makes no sense if I cannot run Postgres or Redis on standard ports because Deis components already have taken these ports.
Changing Deis services' ports could also be a solution.

**Comment by aledbf** [comment]
@ineu about 
> docker ACCEPT rules were inserted before firewall rules, so it didn't block anything.

how you start the firewall?

**Comment by ineu** [comment]
@aledbf I don't start it manually, https://github.com/deis/deis/blob/master/contrib/util/custom-firewall.sh enables iptables-restore automatically.
I'll try creating new cluster, maybe there was some mistake on my side.

**Comment by aledbf** [comment]
> I don't start it manually

just checking for that. 
The thing is `iptables-restore` runs before `network.service` so what you mention about the order is strange. Can you post the output of iptables that you mention (where docker ACCEPT rules appears)?

**Comment by ineu** [comment]
So here's what I see after reboot: https://gist.github.com/ineu/d0d7787fc52fca01d49d

**Comment by ineu** [comment]
BTW, why firewall setup uses icmp-host-prohibited for REJECT rule instead of default port-unreachable?

**Comment by carmstrong** [comment]
> BTW, why firewall setup uses icmp-host-prohibited for REJECT rule instead of default port-unreachable?

Probably because noone noticed before :) I think that's probably a reasonable change if you care to PR it.

**Comment by ineu** [comment]
#3179

**Comment by ineu** [comment]
But what do you guys think about Deis components binding to standard postgres/redis ports thus blocking ability to run separate postgres/redis and expose ports to the host?

**Comment by bacongobbler** [comment]
AFAIK I think we already bind to COREOS_PRIVATE_IPV4, hence why we set $HOST up in the container.

**Comment by ineu** [comment]
@bacongobbler
What I mean is `s/-p 6379:6379/-p $COREOS_PRIVATE_IPV4:6379:6379/` for https://github.com/deis/deis/blob/master/deisctl/units/deis-cache.service#L9. Same for deis-database etc. With current setup it looks like:
```
core@staging-coreos-2 ~ $ docker ps | grep cache
e1605f7d0c81        deis/cache:v1.3.1                        "/app/bin/boot"        14 minutes ago      Up 14 minutes       0.0.0.0:6379->6379/tcp
```
So it is possible to do (from the toolbox):
```
root@staging-coreos-2:~# redis-cli -h 127.0.0.1 ping
PONG
root@staging-coreos-2:~# redis-cli -h 192.168.0.2 ping
PONG
root@staging-coreos-2:~# redis-cli -h 188.40.x.x ping
PONG
```
because it listens not only on 192.168.0.2, but on all interfaces.

Even better would be to make Deis components use other ports instead of default postgres/redis. By this will probably break compatibility, so I doubt it's an option.

**Comment by ineu** [comment]
@aledbf As for the rules order, please take a look at https://gist.github.com/ineu/a980db85da8747c27706

> The thing is iptables-restore runs before network.service so what you mention about the order is strange.

I believe the reason is that old versions of Docker inserts rules to FORWARD table: https://github.com/docker/docker/pull/7003#issuecomment-48854102

This can probably be only fixed by using newest CoreOS with docker 1.5

**Comment by bacongobbler** [comment]
@ineu I believe that's intended. We utilize both the local loopback (127.0.0.1) and the private IP address (192.168...) so having it exposed is necessary. We have measures in place as mentioned above that mitigate the security risks here, so I'm going to close this as something we aren't going to fix unless there's an extremely compelling reason that someone outside the cluster is able to access the database (save for internal applications which is covered by #986).
